### PR TITLE
Add retry operation.

### PR DESCRIPTION
**Pull request by wsargent**
Add a retry operation to the basic httpclient to stop the build failing.

### CODE REVIEW

**Comment by wsargent** [comment]
We'll see if this passes the build tests...

**Comment by richdougherty** [comment]
The current code will always fail when it retries, because it just retries reading from the same closed socket. The socket remains closed so it fails each time. The retry code needs to close the old socket, make a new socket and try again with the new socket.

The best way to do this is to put the retry logic around the entire body of the `makeRequests` method. The start of this method makes a `BasicHttpClient` and the finally block at the end closes it, so the body of the method represents a complete attempt at using a socket.

IMO the retry code should only retry based a whitelist of certain exceptions that are known to be related to reading from a closed socket. We don't want it retrying for all sorts of exceptions, only ones which we're sure are safe.

Here's the stack trace for an exception we want to retry on. You can see how this code is called by `makeRequests`, so placing the retry there would hopefully fix the problem.
```
[error]    Exception while reading response 1 java.lang.RuntimeException: No response 1: EOF reached (BasicHttpClient.scala:225)
[error] play.it.http.BasicHttpClient.readResponse(BasicHttpClient.scala:225)
[error] play.it.http.BasicHttpClient$$anonfun$sendRequest$4.apply(BasicHttpClient.scala:126)
[error] play.it.http.BasicHttpClient$$anonfun$sendRequest$4.apply(BasicHttpClient.scala:124)
[error] play.it.http.BasicHttpClient.sendRequest(BasicHttpClient.scala:124)
[error] play.it.http.BasicHttpClient$$anonfun$1.apply(BasicHttpClient.scala:31)
[error] play.it.http.BasicHttpClient$$anonfun$1.apply(BasicHttpClient.scala:29)
[error] play.it.http.BasicHttpClient$.makeRequests(BasicHttpClient.scala:29)
```

File: framework/src/play-integration-test/src/test/scala/play/it/http/BasicHttpClient.scala [hunk 1]
```diff
@@ -22,33 +24,45 @@ object BasicHttpClient {
...
+    }
+  }
+
+  /** Retries the operation a number of times defined by n. */
+  def retry[T](n: Int)(fn: => T): T = {
+    try {
+      fn
+    } catch {
+      case e if n > 1 =>
+        retry(n - 1)(fn)
```
Comment (jroper) [inline framework/src/play-integration-test/src/test/scala/play/it/http/BasicHttpClient.scala]: I don't think this is the right place to do a retry - many of the tests assume only one request will be made, so retrying if it fails often won't help.  You can retry the whole test by simply changing the 1 here to the number of times you want to retry:

https://github.com/playframework/playframework/blob/master/framework/src/play-integration-test/src/test/scala/play/it/ServerIntegrationSpecification.scala#L28

Comment (richdougherty) [inline framework/src/play-integration-test/src/test/scala/play/it/http/BasicHttpClient.scala]: But that would retry for any failure. I think we really only want to retry for specific IO exceptions.

Comment (wsargent) [inline framework/src/play-integration-test/src/test/scala/play/it/http/BasicHttpClient.scala]: Okay, will restrict to IOException.

**Comment by richdougherty** [comment]
I'm going to merge this now because I think it will improve things.

In the future we might want to add a `withClient { ... }` DSL to `BasicHttpClient` that lets us easily do retries over client connections and work around connection close bugs. At the moment, as @jroper points out, the retry logic only works for tests where multiple requests are made at once. We make multiple requests in a lot of tests but a `withClient { ... }` API would support more tests.

**Pull request merged**